sanic_routing/router.py

```diff
@@ -367,51 +507,68 @@ def routes(self):
             [route for group in self.groups.values() for route in group]
         )
 
-    def optimize(self, src: t.List[Line]) -> None:
-        """
-        Insert NotFound exceptions to be able to bail as quick as possible,
-        and realign lines to proper indentation
-        """
-        offset = 0
-        current = 0
-        insert_at = set()
-        for num, line in enumerate(src):
-            if line.indent < current:
-                if not line.src.startswith("."):
-                    if offset < 0:
-                        offset += 1
-                    else:
-                        offset = 0
-
+    def _optimize(self, node) -> None:
+        warn(
+            "Router AST optimization is an experimental only feature. "
+            "Results may vary from unoptimized code."
+        )
+        if hasattr(node, "body"):
+            for child in node.body:
+                self._optimize(child)
+
+            # concatenate nested single if blocks
+            # EXAMPLE:
+            #      if parts[1] == "foo":
+            #          if num > 3:
+            # BECOMES:
+            #       if parts[1] == 'foo' and num > 3:
+            # Testing has shown that further recursion does not actually
+            # produce any faster results.
+            if self._is_lone_if(node) and self._is_lone_if(node.body[0]):
+                current = node.body[0]
+                nested = node.body[0].body[0]
+
+                values: t.List[t.Any] = []
+                for test in [current.test, nested.test]:
+                    if isinstance(test, ast.Compare):
+                        values.append(test)
+                    elif isinstance(test, ast.BoolOp) and isinstance(
+                        test.op, ast.And
+                    ):
+                        values.extend(test.values)
+                combined = ast.BoolOp(op=ast.And(), values=values)
+
+                current.test = combined
+                current.body = nested.body
+
+            # Look for identical successive if blocks
+            # EXAMPLE:
+            #       if num == 5:
+            #           foo1()
+            #       if num == 5:
+            #           foo2()
+            # BECOMES:
+            #       if num == 5:
+            #           foo1()
+            #           foo2()
             if (
-                line.src.startswith("if")
-                or line.src.startswith("elif")
-                or line.src.startswith("return")
-                or line.src.startswith("basket")
-                or line.src.startswith("try")
+                all(isinstance(child, ast.If) for child in node.body)
+                and len({child.test for child in node.body})
```

Comment:
What is this row supposed to do?

Reply:
:thinking: I think this might be a relic from an earlier test. All `If` objects have a `test` attribute, so as it is written this would never be Falsey. IIRC, at one point I was testing for the number of unique comparisons. I think I can remove this safely.

Reply:
I am commenting it out for now, but will leave the code in there with a note for something to come back to next round.